lib/class-wp-theme-json.php

```diff
@@ -295,75 +272,53 @@ class WP_Theme_JSON {
 	 * @param array $theme_json A structure that follows the theme.json schema.
 	 */
 	public function __construct( $theme_json = array() ) {
-		$this->theme_json = array();
-
-		if ( ! is_array( $theme_json ) ) {
-			return;
-		}
-
-		// Remove top-level keys that aren't present in the schema.
-		$this->theme_json = array_intersect_key( $theme_json, self::SCHEMA );
+		$valid_block_names = array_keys( self::get_blocks_metadata() );
+		$this->theme_json  = self::sanitize( $theme_json, $valid_block_names );
+	}
 
-		$block_metadata = self::get_blocks_metadata();
-		foreach ( array( 'settings', 'styles' ) as $subtree ) {
-			// Remove settings & styles subtrees if they aren't arrays.
-			if ( isset( $this->theme_json[ $subtree ] ) && ! is_array( $this->theme_json[ $subtree ] ) ) {
-				unset( $this->theme_json[ $subtree ] );
-			}
+	/**
+	 * Sanitizes the input according to the schemas.
+	 *
+	 * @param array $input Structure to sanitize.
+	 * @param array $valid_block_names List of valid block names.
+	 *
+	 * @return array The sanitized output.
+	 */
+	private static function sanitize( $input, $valid_block_names ) {
+		$output = array();
 
-			// Remove block selectors subtrees declared within settings & styles if that aren't registered.
-			if ( isset( $this->theme_json[ $subtree ] ) ) {
-				$this->theme_json[ $subtree ] = array_intersect_key( $this->theme_json[ $subtree ], $block_metadata );
-			}
+		if ( ! is_array( $input ) ) {
+			return $output;
 		}
 
-		foreach ( $block_metadata as $block_selector => $metadata ) {
-			if ( isset( $this->theme_json['styles'][ $block_selector ] ) ) {
-				// Remove the block selector subtree if it's not an array.
-				if ( ! is_array( $this->theme_json['styles'][ $block_selector ] ) ) {
-					unset( $this->theme_json['styles'][ $block_selector ] );
-					continue;
-				}
+		$output = array_intersect_key( $input, array_flip( self::VALID_TOP_LEVEL_KEYS ) );
 
-				$styles_schema                                 = self::SCHEMA['styles'];
-				$this->theme_json['styles'][ $block_selector ] = self::remove_keys_not_in_schema(
-					$this->theme_json['styles'][ $block_selector ],
-					$styles_schema
-				);
+		$schema = array();
+		foreach ( $valid_block_names as $block_name ) {
+			$schema['styles'][ $block_name ]   = self::VALID_STYLES;
+			$schema['settings'][ $block_name ] = self::VALID_SETTINGS;
+		}
```

Comment:
For each `$block_name`, we're creating an identical copy of `self::VALID_STYLES` and `self::VALID_SETTINGS`, respectively. Would it be worth considering having just

```suggestion
		$schema['styles']   = self::VALID_STYLES;
		$schema['settings'] = self::VALID_SETTINGS;
```

and changing `remove_keys_not_in_schema()` accordingly?

Reply:
One of the struggles of this class has been how to make it adaptable to theme.json shape changes. The idea I explored in the latest rounds is to separate the logic that iterates over the tree from the logic to do something with the tree or a particular set of nodes. A visitor pattern of sorts. The hope is that, by centralizing the iteration logic, changes can be made easily in the future.

For example, in this case, `remove_keys_not_in_schema` is agnostic about the shape of theme.json, it just cares about removing nodes in the given tree that don't follow the given schema. The same logic applies to some other methods in this class after the latest changes landed in master: most of them become "visitors" of nodes, they know how to operate on a node or set of nodes (take styles out of them, process presets, etc) but are theme.json-shape agnostic. Then there're a couple of methods that are focused on building the paths to iterate over, so those are the only ones that need updating when theme.json changes.